File: 程序员面试金典/8.2迷路的机器人/pathWithObstacles.cpp
class Solution {
public:
    vector<vector<int>> pathWithObstacles(vector<vector<int>>& obstacleGrid) {
        vector<vector<int>> ans;
        if(!obstacleGrid.size() || !obstacleGrid[0].size())
            return ans;
        int r = obstacleGrid.size(), c = obstacleGrid[0].size();
        if(obstacleGrid[r - 1][c - 1])
            return ans;
        vector<vector<int>> dp(r, vector<int>(c, 0));
        for(int i = 0; i < r; ++i) {
            if(obstacleGrid[i][0] == 1)
                break;
            dp[i][0] = 1;
        }
        for(int j = 0; j < c; ++j) {
            if(obstacleGrid[0][j] == 1)
                break;
            dp[0][j] = 1;
        }
        for(int i = 1; i < r; ++i)
            for(int j = 1; j < c; ++j)
                dp[i][j] = !obstacleGrid[i][j] ? max(dp[i - 1][j], dp[i][j - 1]) : 0;
        if(!dp[r - 1][c - 1])
            return ans;
        int i = r - 1, j = c - 1;
        while(i >= 0 && j >= 0) {
            ans.push_back({i, j});
            if (i - 1 >= 0 && dp[i - 1][j])
                --i;
            else
                --j;
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: 程序员面试金典/8.2/pathWithObstacles.py
class Solution:
    def pathWithObstacles(self, obstacleGrid: List[List[int]]) -> List[List[int]]:
        if obstacleGrid[0][0] == 1 or obstacleGrid[-1][-1] == 1:
            return []
        path = [[0, 0]]
        r, c = len(obstacleGrid), len(obstacleGrid[0])
        visited = [[False] * c for i in range(r)]
        visited[0][0] = True
        return path if self.dfs(obstacleGrid, r, c, 0, 0, path, visited) else []
    
    def dfs(self, grid, r, c, x, y, path, visited):
        if x == r - 1 and y == c - 1:
            return True
        if y < c - 1 and grid[x][y + 1] == 0 and not visited[x][y + 1]:
            path.append([x, y + 1])
            visited[x][y + 1] = True
            if self.dfs(grid, r, c, x, y + 1, path, visited):
                return True
            path.pop()
        if x < r - 1 and grid[x + 1][y] == 0 and not visited[x + 1][y]:
            path.append([x + 1, y])
            visited[x + 1][y] = True
            if self.dfs(grid, r, c, x + 1, y, path, visited):
                return True
            path.pop()
        return False
